les/utils/limiter.go

```diff
@@ -0,0 +1,405 @@
+
+// add adds the node queue to the address group. It is the caller's responsibility to
+// add the address group to the address map and the address selector if it wasn't
+// there before.
+func (ag *addressGroup) add(nq *nodeQueue) {
+	if nq.groupIndex != -1 {
+		panic("added node queue is already in an address group")
+	}
+	l := len(ag.nodes)
+	nq.groupIndex = l
+	ag.nodes = append(ag.nodes, nq)
+	ag.sumFlatWeight += nq.flatWeight
+	ag.groupWeight = ag.sumFlatWeight / uint64(l+1)
+	ag.nodeSelect.Update(ag.nodes[l])
+}
+
+// update updates the selection weight of the node queue inside the address group.
+// It is the caller's responsibility to update the group's selection weight in the
+// address selector.
+func (ag *addressGroup) update(nq *nodeQueue, weight uint64) {
+	if nq.groupIndex == -1 || nq.groupIndex >= len(ag.nodes) || ag.nodes[nq.groupIndex] != nq {
+		panic("updated node queue is not in this address group")
+	}
+	ag.sumFlatWeight += weight - nq.flatWeight
+	nq.flatWeight = weight
+	ag.groupWeight = ag.sumFlatWeight / uint64(len(ag.nodes))
+	ag.nodeSelect.Update(nq)
+}
+
+// remove removes the node queue from the address group. It is the caller's responsibility
+// to remove the address group from the address map if it is empty.
+func (ag *addressGroup) remove(nq *nodeQueue) {
+	if nq.groupIndex == -1 || nq.groupIndex >= len(ag.nodes) || ag.nodes[nq.groupIndex] != nq {
+		panic("removed node queue is not in this address group")
+	}
+
+	l := len(ag.nodes) - 1
+	if nq.groupIndex != l {
+		ag.nodes[nq.groupIndex] = ag.nodes[l]
+		ag.nodes[nq.groupIndex].groupIndex = nq.groupIndex
+	}
+	nq.groupIndex = -1
+	ag.nodes = ag.nodes[:l]
+	ag.sumFlatWeight -= nq.flatWeight
+	if l >= 1 {
+		ag.groupWeight = ag.sumFlatWeight / uint64(l)
+	} else {
+		ag.groupWeight = 0
+	}
+	ag.nodeSelect.Remove(nq)
+}
+
+// choose selects one of the node queues belonging to the address group
+func (ag *addressGroup) choose() *nodeQueue {
+	return ag.nodeSelect.Choose().(*nodeQueue)
+}
+
+// NewLimiter creates a new Limiter
+func NewLimiter(sumCostLimit uint) *Limiter {
+	l := &Limiter{
+		addressSelect: NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*addressGroup).groupWeight }),
+		valueSelect:   NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*nodeQueue).valueWeight }),
+		nodes:         make(map[enode.ID]*nodeQueue),
+		addresses:     make(map[string]*addressGroup),
+		sumCostLimit:  sumCostLimit,
+	}
+	l.cond = sync.NewCond(&l.lock)
+	go l.processLoop()
+	return l
+}
+
+// selectionWeights calculates the selection weights of a node for both the address and
+// the value selector. The selection weight depends on the next request cost or the
+// summed cost of recently dropped requests.
+func (l *Limiter) selectionWeights(reqCost uint, value float64) (flatWeight, valueWeight uint64) {
+	if value > l.maxValue {
+		l.maxValue = value
+	}
+	if value > 0 {
```

Comment:
No, negative values are not supported, they wouldn't make any sense. The check is there to avoid a 0/0 division. Checking either `value` or `maxValue` for `> 0` or `!= 0` would work fine here.